packages/website/src/components/ASTTsViewer.tsx

```diff
@@ -0,0 +1,149 @@
+import React, { useCallback, useEffect, useState } from 'react';
+
+import ASTViewer from './ast/ASTViewer';
+import { isRecord } from './ast/utils';
+import type {
+  ASTViewerBaseProps,
+  SelectedRange,
+  SelectedPosition,
+} from './ast/types';
+import type { Node, SourceFile } from 'typescript';
+
+export interface ASTTsViewerProps extends ASTViewerBaseProps {
+  readonly version: string;
+}
+
+function extractEnum(
+  obj: Record<number, string | number>,
+): Record<number, string> {
+  const result: Record<number, string> = {};
+  const keys = Object.entries(obj);
+  for (const [name, value] of keys) {
+    if (typeof value === 'number') {
+      if (!(value in result)) {
+        result[value] = name;
+      }
+    }
+  }
+  return result;
+}
+
+function isTsNode(value: unknown): value is Node {
+  return isRecord(value) && typeof value.kind === 'number';
+}
+
+function getFlagNamesFromEnum(
+  allFlags: Record<number, string>,
+  flags: number,
+  prefix: string,
+): string[] {
+  return Object.entries(allFlags)
+    .filter(([f, _]) => (Number(f) & flags) !== 0)
+    .map(([_, name]) => `${prefix}.${name}`);
+}
+
+export function getLineAndCharacterFor(
+  pos: number,
+  ast: SourceFile,
+): SelectedPosition {
+  const loc = ast.getLineAndCharacterOfPosition(pos);
+  return {
+    line: loc.line + 1,
+    column: loc.character,
+  };
+}
+
+export function getLocFor(
+  start: number,
+  end: number,
+  ast: SourceFile,
+): SelectedRange {
+  return {
+    start: getLineAndCharacterFor(start, ast),
+    end: getLineAndCharacterFor(end, ast),
+  };
+}
+
+export const propsToFilter = [
+  'parent',
+  'jsDoc',
+  'lineMap',
+  'externalModuleIndicator',
+  'bindDiagnostics',
+  'transformFlags',
+  'resolvedModules',
+  'imports',
+];
+
+export default function ASTTsViewer(props: ASTTsViewerProps): JSX.Element {
+  const [syntaxKind, setSyntaxKind] = useState<Record<number, string>>({});
+  const [nodeFlags, setNodeFlags] = useState<Record<number, string>>({});
+  const [tokenFlags, setTokenFlags] = useState<Record<number, string>>({});
+  const [modifierFlags, setModifierFlags] = useState<Record<number, string>>(
+    {},
+  );
+
+  useEffect(() => {
+    setSyntaxKind(extractEnum(window.ts.SyntaxKind));
+    setNodeFlags(extractEnum(window.ts.NodeFlags));
+    setTokenFlags(extractEnum(window.ts.TokenFlags));
+    setModifierFlags(extractEnum(window.ts.ModifierFlags));
+  }, [props.version]);
```

Comment:
question: does this need to be done in an effect or could it be done inline?
I'm assuming you've done it in an effect as TS may not have loaded yet?

Reply:
ts is going to be always loaded, at this point, i did that in effect for reactivity and to cache those values

I'm trying to minimize how much operations we have to do render tree